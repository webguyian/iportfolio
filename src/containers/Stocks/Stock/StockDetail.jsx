import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import { formatPrice, getPercentage } from 'modules/stocks/helpers';

import Button from 'components/Button/Button';
import Text from 'components/Text/Text';

import StockChart from 'containers/Stocks/StockChart/StockChart';
import StockNews from 'containers/Stocks/StockNews/StockNews';
import StockTicker from 'containers/Stocks/StockTicker/StockTicker';

const StockDetail = props => {
  const { onAdd, onClick, onClose, stock, stocks } = props;
  const baseClass = 'stock-detail';
  const differenceClass = `${baseClass}-difference`;

  if (!stock) {
    return null;
  }

  const stockSymbols = stocks.map(s => s.displaySymbol);
  const showAdd = !stockSymbols.includes(stock.displaySymbol);
  const modifierClass =
    stock.price > stock.previousPrice
      ? `${differenceClass}--up`
      : `${differenceClass}--down`;

  return (
    <Fragment>
      <StockTicker stocks={stocks} onClick={onClick} />
      <div className={baseClass}>
        <header className={`${baseClass}-header`}>
          <Text className={`${baseClass}-symbol`} modifier="bold">
            {stock.symbol}
          </Text>
          <Text className={`${baseClass}-description`}>
            {stock.description}
          </Text>
          <Button
            className={`${baseClass}-close-btn`}
            icon="times-circle"
            size="2x"
            onClick={onClose}
          >
            Close detail view
          </Button>
        </header>
        <div className={`${baseClass}-content`}>
          <div className={`${baseClass}-value`}>
            <Text className={`${baseClass}-price`}>
              {formatPrice(stock.price)}
            </Text>
            <Text className={cx(differenceClass, modifierClass)}>
              {getPercentage(stock.price, stock.previousPrice)}
            </Text>
            {showAdd && (
              <Button className={`${baseClass}-add-btn`} onClick={onAdd}>
                Add
              </Button>
            )}
          </div>
          <StockChart stock={stock} />
          <StockNews category={stock.symbol} displayCount={5} />
        </div>
      </div>
    </Fragment>
  );
};

StockDetail.propTypes = {
  onAdd: PropTypes.func.isRequired,
  onClick: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired,
  stock: PropTypes.object,
  stocks: PropTypes.array
};

export default StockDetail;
