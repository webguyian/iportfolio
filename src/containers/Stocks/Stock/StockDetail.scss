.stock-detail {
  z-index: 9;
  position: absolute;
  top: 65px;
  left: 0;
  height: 100%;
  border-top-left-radius: 10px;
  border-top-right-radius: 10px;
  background-color: $black-dark;

  &-header {
    display: flex;
    position: relative;
    align-items: flex-end;
    padding: $gutter;
    border-bottom: 1px solid $gray-medium;
  }

  &-value {
    display: flex;
    align-items: flex-end;
    padding: $gutter-s $gutter;
  }

  &-symbol {
    margin-right: $gutter-s;
    font-size: 25px;
    line-height: 1;
  }

  &-description {
    color: $gray;
    font-size: $font-size-s;
  }

  &-close-btn {
    margin-left: auto;
    color: $gray;
  }

  &-price,
  &-difference {
    font-weight: $font-weight-bold;
  }

  &-price {
    margin-right: $gutter;
  }

  &-difference {
    font-size: $font-size-s;

    &--down {
      color: $red;
    }

    &--up {
      color: $green-light;
    }
  }

  &-add-btn {
    @include type-body($font-size-s);
    margin-left: auto;
    padding: 4px 12px;
    border-radius: 10px;
    background: $blue-light;
    color: $black-dark;
  }

  &-content {
    max-height: 650px;
    overflow-y: auto;
  }

  &-chart {
    min-height: 300px;
    margin-bottom: $gutter;
    padding: 0 $gutter;
    background-color: $gray;
  }

  .stock-news {
    position: static;
    width: 100%;
    margin: 0;
    transform: none;
    background: none;

    &-header {
      padding-bottom: 0;
      border-bottom: 0;
    }
  }
}
