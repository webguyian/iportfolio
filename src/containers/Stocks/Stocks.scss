.stocks-app {
  @include stretch-column-nowrap;
  position: relative;
  padding: $gutter;
  transition: transform 0.3s ease;

  &--with-search {
    padding-top: 70px;
    transform: translateY(-70px);

    .stocks-app-header {
      display: none;
    }

    .stocks-app-list {
      max-height: none;
    }

    .ui-search-field {
      margin: $gutter-s 4px;

      > .ui-btn--anchor {
        display: block; // Show cancel button
      }
    }
  }

  &--with-detail {
    padding-top: $gutter;
    transform: none;

    > .stock-news {
      transform: translateY(110%);
    }
  }

  &-header {
    display: flex;
    align-items: flex-start;
    justify-content: space-between;
    padding: 0 $gutter-s;
  }

  &-title,
  &-date {
    font-size: $font-size-l;
  }

  &-date {
    color: $gray;
  }

  &-edit-btn {
    @include type-body;
    padding: 4px $gutter;
    color: $blue-light;
  }

  &-list {
    @include unstyled-list;
    max-height: 600px;
    overflow-y: auto;
  }
}
