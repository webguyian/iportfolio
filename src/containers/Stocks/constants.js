export const initialStocks = [
  { symbol: 'AAPL', displaySymbol: 'AAPL', description: 'APPLE INC' },
  { symbol: 'AMZN', displaySymbol: 'AMZN', description: 'AMAZON.COM INC' },
  { symbol: 'FB', displaySymbol: 'FB', description: 'FACEBOOK INC-CLASS A' },
  { symbol: 'GOOG', displaySymbol: 'GOOG', description: 'ALPHABET INC-CL C' },
  { symbol: 'MSFT', displaySymbol: 'MSFT', description: 'MICROSOFT CORP' }
];

export const API_CANDLESTICK = 'https://finnhub.io/api/v1/stock/candle';

export const API_NEWS = 'https://finnhub.io/api/v1/news';

export const API_QUOTE = 'https://finnhub.io/api/v1/quote';

export const API_TOKEN = 'bojsah7rh5rcji5m4udg';

export const API_EXCHANGE = `https://finnhub.io/api/v1/stock/symbol?exchange=US&token=${API_TOKEN}`;
