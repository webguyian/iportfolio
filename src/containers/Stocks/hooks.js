import { useEffect, useState } from 'react';

import { useStorageCache } from 'hooks';
import { API_EXCHANGE, API_TOKEN } from 'containers/Stocks/constants';
import {
  getCandlestickEndpoint,
  getCandlestickEndpoints,
  getStockEndpoint,
  getStockNewsEndpoint,
  isExpired,
  isExpiredNews,
  updateChartData,
  rehydrateChartData
} from 'containers/Stocks/helpers';

export const useFetch = endpoint => {
  const [data, setData] = useState(null);
  const fetchData = async () => {
    const response = await fetch(endpoint);
    const result = await response.json();

    setData(result);
  };

  useEffect(() => {
    if (!endpoint) {
      return;
    }

    fetchData();
  }, [endpoint]);

  return data;
};

export const useFetchAll = urls => {
  const [responses, setResponses] = useState([]);

  useEffect(() => {
    const fetchData = async () => {
      const results = await Promise.all(
        urls.map(async url => {
          const response = await fetch(url);

          return response.json();
        })
      );

      setResponses(results);
    };

    if (urls.length) {
      fetchData();
    }
  }, [urls]);

  return responses;
};

export const useSearch = () => {
  const [value, setValue] = useState('');
  const [searchTerm, setSearchTerm] = useState('');
  const [searchFocus, setSearchFocus] = useState(false);
  const hasSearch = Boolean(searchFocus || searchTerm.length);

  const handleCancel = () => {
    setValue('');
    setSearchTerm('');
    setSearchFocus(false);
  };

  const handleChange = event => {
    setValue(event.target.value);
  };

  const handleFocus = () => {
    setSearchFocus(focus => !focus);
  };

  const searchHandlers = {
    onBlur: handleFocus,
    onCancel: handleCancel,
    onChange: handleChange,
    onFocus: handleFocus,
    value
  };

  useEffect(() => {
    if (value) {
      setSearchTerm(value.toUpperCase().trim());
    } else {
      setSearchTerm('');
    }
  }, [value]);

  return [searchTerm, searchHandlers, hasSearch];
};

export const useStockSearch = (allStocks, searchTerm) => {
  const [filteredStocks, setFilteredStocks] = useState(allStocks);

  useEffect(() => {
    if (searchTerm.length > 1) {
      const filtered = allStocks.filter(stock => {
        return (
          stock.symbol.indexOf(searchTerm) === 0 ||
          stock.description.indexOf(searchTerm) === 0
        );
      });

      setFilteredStocks(filtered);
    } else if (!searchTerm) {
      setFilteredStocks(allStocks);
    }
  }, [searchTerm]);

  return filteredStocks;
};

export const useStocks = initialStocks => {
  const [endpoints, setEndpoints] = useState([]);
  const [stocks, setStocks] = useState([]);
  const [allStocks, setAllStocks] = useState([]);
  const [activeStocks, setActiveStocks] = useState([]);
  const exchangeResponse = useFetch(API_EXCHANGE);
  const response = useFetchAll(endpoints);
  const cache = useStorageCache(
    'stocks',
    { stocks },
    stockData => !stockData.stocks.length
  );

  useEffect(() => {
    const cachedStocks = cache && cache.stocks;

    if (cachedStocks && cachedStocks.length) {
      setStocks(cachedStocks);
      setActiveStocks(cachedStocks);
    } else {
      setStocks(initialStocks);
      setActiveStocks(initialStocks);
    }
  }, []);

  useEffect(() => {
    if (!exchangeResponse) {
      // Exit early if no response
      return;
    }

    const symbols = activeStocks.map(stock => stock.symbol);
    const currentStocks = exchangeResponse.filter(stock =>
      symbols.includes(stock.symbol)
    );

    setEndpoints(currentStocks.map(getStockEndpoint));
    setAllStocks(exchangeResponse);
  }, [exchangeResponse]);

  useEffect(() => {
    if (activeStocks.length) {
      setEndpoints(activeStocks.map(getStockEndpoint));
    }
  }, [activeStocks]);

  useEffect(() => {
    if (!response.length || !activeStocks.length) {
      return;
    }

    const updatedStocks = response.map((result, index) => {
      const stock = activeStocks[index];

      return {
        ...result,
        ...stock,
        price: result.c,
        previousPrice: result.pc
      };
    });

    setStocks(updatedStocks);
  }, [response, activeStocks]);

  return [stocks, setStocks, allStocks];
};

export const useStockDetail = () => {
  const [endpoint, setEndpoint] = useState('');
  const [activeStock, setActiveStock] = useState(null);
  const [stock, setStockDetail] = useState(activeStock);
  const response = useFetch(endpoint);
  const getStock = symbol =>
    `https://finnhub.io/api/v1/quote?symbol=${symbol}&token=${API_TOKEN}`;

  useEffect(() => {
    setStockDetail(activeStock);

    if (!activeStock) {
      return;
    }

    if (!activeStock.price) {
      setEndpoint(getStock(activeStock.symbol));
    }
  }, [activeStock]);

  useEffect(() => {
    if (response) {
      setStockDetail({
        ...activeStock,
        ...response,
        price: response.c,
        previousPrice: response.pc
      });
      setEndpoint('');
    }
  }, [response]);

  return [stock, setActiveStock];
};

export const useStockNews = category => {
  const [endpoint, setEndpoint] = useState('');
  const [news, setNews] = useState([]);
  const response = useFetch(endpoint);
  const cache = useStorageCache(
    'stock-news',
    { news: response },
    res => !res.news.length
  );

  useEffect(() => {
    const cachedNews = cache && cache.news;

    if (cachedNews && cachedNews.length) {
      setNews(cachedNews);
      const expired = isExpiredNews(cachedNews[0].datetime);

      if (!expired && !category) {
        // Exit early when cache is not expired and no category
        return;
      }
    }

    setEndpoint(getStockNewsEndpoint(category));
  }, [category]);

  useEffect(() => {
    if (response && response.length) {
      setNews(response);
    }
  }, [response]);

  return news;
};

export const useStockChart = (symbol, range) => {
  const [endpoint, setEndpoint] = useState();
  const [updatedResponse, setResponse] = useState(null);
  const response = useFetch(endpoint);

  useEffect(() => {
    const updatedEndpoint = getCandlestickEndpoint(symbol, range);

    setEndpoint(updatedEndpoint);
  }, [symbol, range]);

  useEffect(() => {
    if (response && response.s === 'ok') {
      const data = updateChartData(response, range);

      setResponse(data);
    }
  }, [response]);

  return updatedResponse;
};

export const useStockTicker = (stocks, range = '1D') => {
  const [endpoints, setEndpoints] = useState([]);
  const [updatedResponse, setResponse] = useState([]);
  const response = useFetchAll(endpoints);
  const cache = useStorageCache(
    'stock-charts',
    { charts: updatedResponse },
    res => !res.charts.length
  );

  useEffect(() => {
    const cachedCharts = cache && cache.charts;

    if (cachedCharts && cachedCharts.length) {
      const updated = cachedCharts.map(rehydrateChartData);
      const [head] = updated;
      const expired = isExpired(head.chartData.endDate);

      setResponse(updated);

      if (!expired) {
        // Exit early when cache is not expired
        return;
      }
    }

    // Trigger fetch for candlestick data
    setEndpoints(getCandlestickEndpoints(stocks, range));
  }, [stocks.length]);

  useEffect(() => {
    if (response.length) {
      const updated = response.map((res, index) => {
        if (res.s !== 'ok') {
          return null;
        }

        const stock = stocks[index];
        const chartData = updateChartData(res, range);

        return {
          ...stock,
          chartData
        };
      });

      if (updated[0] !== null) {
        setResponse(updated);
      }
    }
  }, [response]);

  return updatedResponse;
};
