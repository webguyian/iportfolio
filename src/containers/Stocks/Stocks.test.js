import React from 'react';
import { act, create } from 'react-test-renderer';

import { initialStocks } from 'containers/Stocks/constants';
import * as hooks from 'containers/Stocks/hooks';
import StockDetail from 'containers/Stocks/Stock/StockDetail';
import Stocks from './Stocks';

jest.mock('containers/Stocks/StockChart/StockChart', () => 'StockChart');
jest.mock('containers/Stocks/StockNews/StockNews', () => 'StockNews');
jest.mock('containers/Stocks/StockTicker/StockTicker', () => 'StockTicker');

describe('<Stocks />', () => {
  const useStocksHook = hooks.useStocks;
  const useSearchHook = hooks.useSearch;
  const useStockDetailHook = hooks.useStockDetail;

  hooks.useStocks = jest.fn();
  hooks.useSearch = jest.fn();
  hooks.useStockDetail = jest.fn();

  const searchHandlers = {
    onBlur: jest.fn(),
    onCancel: jest.fn(),
    onChange: jest.fn(),
    onFocus: jest.fn(),
    value: 'aa'
  };

  beforeEach(() => {
    jest.clearAllMocks();
    // Reset to original implementation before each test
    hooks.useStocks.mockImplementation(useStocksHook);
    hooks.useSearch.mockImplementation(useSearchHook);
    hooks.useStockDetail.mockImplementation(useStockDetailHook);
  });

  it('renders correctly', () => {
    hooks.useStocks.mockReturnValue([initialStocks, jest.fn(), initialStocks]);

    const component = create(<Stocks />);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with search term', () => {
    hooks.useStocks.mockReturnValue([initialStocks, jest.fn(), initialStocks]);
    hooks.useSearch.mockReturnValue(['aa', searchHandlers, true]);
    const component = create(<Stocks />);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with stock detail', () => {
    const activeStock = {
      symbol: 'AAPL',
      displaySymbol: 'AAPL',
      description: 'APPLE INC'
    };

    hooks.useStocks.mockReturnValue([initialStocks, jest.fn(), initialStocks]);
    hooks.useSearch.mockReturnValue(['aa', searchHandlers, true]);
    hooks.useStockDetail.mockReturnValue([activeStock, jest.fn()]);
    const component = create(<Stocks />);

    expect(component).toMatchSnapshot();
  });

  it('handles adding stocks', () => {
    const activeStock = {
      symbol: 'AAPL',
      displaySymbol: 'AAPL',
      description: 'APPLE INC'
    };
    let activeStocks = initialStocks.slice(0);
    const setStocks = stocks => {
      activeStocks = stocks;
    };

    hooks.useStocks.mockReturnValue([activeStocks, setStocks, initialStocks]);
    hooks.useSearch.mockReturnValue(['aa', searchHandlers, true]);
    hooks.useStockDetail.mockReturnValue([activeStock, jest.fn()]);
    const component = create(<Stocks />);
    const detail = component.root.findByType(StockDetail);

    expect(activeStocks.length).toEqual(5);
    expect(searchHandlers.onCancel).not.toHaveBeenCalled();
    act(() => {
      detail.props.onAdd();
    });
    expect(searchHandlers.onCancel).toHaveBeenCalled();
    expect(activeStocks.length).toEqual(6);
  });

  it('handles deleting stocks', () => {
    hooks.useStocks.mockReturnValue([initialStocks, jest.fn(), initialStocks]);
    const component = create(<Stocks />);
    const stock = component.root.findByProps({ symbol: 'AAPL' });

    act(() => {
      stock.props.onDelete();
    });

    expect(component).toMatchSnapshot();
  });

  it('handles toggling edit', () => {
    hooks.useStocks.mockReturnValue([initialStocks, jest.fn(), initialStocks]);
    const component = create(<Stocks />);
    const [button] = component.root.findAllByProps({
      className: 'ui-btn--anchor'
    });

    act(() => {
      button.props.onClick();
    });

    expect(component).toMatchSnapshot();
  });
});
