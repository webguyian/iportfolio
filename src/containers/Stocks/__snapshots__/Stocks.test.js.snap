// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stocks /> handles deleting stocks 1`] = `
<div
  className="stocks-app"
>
  <header
    className="stocks-app-header"
  >
    <div
      className="stocks-app-header-content"
    >
      <h1
        className="ui-text ui-text--display stocks-app-title"
      >
        Stocks
      </h1>
      <h2
        className="ui-text ui-text--display stocks-app-date"
      >
        October 1
      </h2>
    </div>
    <button
      className="ui-btn ui-btn--anchor"
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </header>
  <form
    className="ui-search-field"
  >
    <i
      className="ui-icon ui-icon-search fas fa-search"
    />
    <input
      aria-label="Search"
      className="ui-search-field-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Search"
      value=""
    />
    <button
      className="ui-btn ui-btn--anchor"
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
  </form>
  <ul
    className="stocks-app-list"
  >
    <li
      className="stock"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            AAPL
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            APPLE INC
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
    <li
      className="stock"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            AMZN
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            AMAZON.COM INC
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
    <li
      className="stock"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            FB
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            FACEBOOK INC-CLASS A
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
    <li
      className="stock"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            GOOG
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            ALPHABET INC-CL C
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
    <li
      className="stock"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            MSFT
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            MICROSOFT CORP
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
  </ul>
  <StockNews
    draggable={true}
  />
</div>
`;

exports[`<Stocks /> handles toggling edit 1`] = `
<div
  className="stocks-app stocks-app--with-detail"
>
  <header
    className="stocks-app-header"
  >
    <div
      className="stocks-app-header-content"
    >
      <h1
        className="ui-text ui-text--display stocks-app-title"
      >
        Stocks
      </h1>
      <h2
        className="ui-text ui-text--display stocks-app-date"
      >
        October 1
      </h2>
    </div>
    <button
      className="ui-btn ui-btn--anchor"
      onClick={[Function]}
      type="button"
    >
      Done
    </button>
  </header>
  <form
    className="ui-search-field"
  >
    <i
      className="ui-icon ui-icon-search fas fa-search"
    />
    <input
      aria-label="Search"
      className="ui-search-field-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Search"
      value=""
    />
    <button
      className="ui-btn ui-btn--anchor"
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
  </form>
  <ul
    className="stocks-app-list"
  >
    <li
      className="stock stock--swiped"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            AAPL
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            APPLE INC
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
    <li
      className="stock stock--swiped"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            AMZN
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            AMAZON.COM INC
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
    <li
      className="stock stock--swiped"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            FB
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            FACEBOOK INC-CLASS A
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
    <li
      className="stock stock--swiped"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            GOOG
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            ALPHABET INC-CL C
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
    <li
      className="stock stock--swiped"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            MSFT
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            MICROSOFT CORP
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
  </ul>
  <StockNews
    draggable={true}
  />
</div>
`;

exports[`<Stocks /> renders correctly 1`] = `
<div
  className="stocks-app"
>
  <header
    className="stocks-app-header"
  >
    <div
      className="stocks-app-header-content"
    >
      <h1
        className="ui-text ui-text--display stocks-app-title"
      >
        Stocks
      </h1>
      <h2
        className="ui-text ui-text--display stocks-app-date"
      >
        October 1
      </h2>
    </div>
    <button
      className="ui-btn ui-btn--anchor"
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </header>
  <form
    className="ui-search-field"
  >
    <i
      className="ui-icon ui-icon-search fas fa-search"
    />
    <input
      aria-label="Search"
      className="ui-search-field-input"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="Search"
      value=""
    />
    <button
      className="ui-btn ui-btn--anchor"
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
  </form>
  <ul
    className="stocks-app-list"
  >
    <li
      className="stock"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            AAPL
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            APPLE INC
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
    <li
      className="stock"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            AMZN
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            AMAZON.COM INC
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
    <li
      className="stock"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            FB
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            FACEBOOK INC-CLASS A
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
    <li
      className="stock"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            GOOG
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            ALPHABET INC-CL C
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
    <li
      className="stock"
      onMouseDown={[Function]}
    >
      <button
        className="ui-btn stock-btn"
        onClick={[Function]}
        type="button"
      >
        <div
          className="stock-name"
        >
          <span
            className="ui-text ui-text--body ui-text--bold stock-symbol"
          >
            MSFT
          </span>
          <span
            className="ui-text ui-text--body stock-description"
          >
            MICROSOFT CORP
          </span>
        </div>
      </button>
      <button
        className="ui-btn ui-btn--delete"
        onClick={[Function]}
        type="button"
      >
        Delete
      </button>
    </li>
  </ul>
  <StockNews
    draggable={true}
  />
</div>
`;

exports[`<Stocks /> renders correctly with search term 1`] = `
<div
  className="stocks-app stocks-app--with-search"
>
  <header
    className="stocks-app-header"
  >
    <div
      className="stocks-app-header-content"
    >
      <h1
        className="ui-text ui-text--display stocks-app-title"
      >
        Stocks
      </h1>
      <h2
        className="ui-text ui-text--display stocks-app-date"
      >
        October 1
      </h2>
    </div>
    <button
      className="ui-btn ui-btn--anchor"
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </header>
  <form
    className="ui-search-field"
  >
    <i
      className="ui-icon ui-icon-search fas fa-search"
    />
    <input
      aria-label="Search"
      className="ui-search-field-input"
      onBlur={[MockFunction]}
      onChange={[MockFunction]}
      onFocus={[MockFunction]}
      placeholder="Search"
      value="aa"
    />
    <button
      className="ui-btn ui-btn--anchor"
      onClick={[MockFunction]}
      type="button"
    >
      Cancel
    </button>
  </form>
  <ul
    className="stocks-app-list"
  />
  <StockNews
    draggable={true}
  />
</div>
`;

exports[`<Stocks /> renders correctly with stock detail 1`] = `
<div
  className="stocks-app stocks-app--with-detail stocks-app--with-search"
>
  <header
    className="stocks-app-header"
  >
    <div
      className="stocks-app-header-content"
    >
      <h1
        className="ui-text ui-text--display stocks-app-title"
      >
        Stocks
      </h1>
      <h2
        className="ui-text ui-text--display stocks-app-date"
      >
        October 1
      </h2>
    </div>
    <button
      className="ui-btn ui-btn--anchor"
      onClick={[Function]}
      type="button"
    >
      Edit
    </button>
  </header>
  <form
    className="ui-search-field"
  >
    <i
      className="ui-icon ui-icon-search fas fa-search"
    />
    <input
      aria-label="Search"
      className="ui-search-field-input"
      onBlur={[MockFunction]}
      onChange={[MockFunction]}
      onFocus={[MockFunction]}
      placeholder="Search"
      value="aa"
    />
    <button
      className="ui-btn ui-btn--anchor"
      onClick={[MockFunction]}
      type="button"
    >
      Cancel
    </button>
  </form>
  <ul
    className="stocks-app-list"
  />
  <StockTicker
    onClick={[Function]}
    stocks={
      Array [
        Object {
          "description": "APPLE INC",
          "displaySymbol": "AAPL",
          "symbol": "AAPL",
        },
        Object {
          "description": "AMAZON.COM INC",
          "displaySymbol": "AMZN",
          "symbol": "AMZN",
        },
        Object {
          "description": "FACEBOOK INC-CLASS A",
          "displaySymbol": "FB",
          "symbol": "FB",
        },
        Object {
          "description": "ALPHABET INC-CL C",
          "displaySymbol": "GOOG",
          "symbol": "GOOG",
        },
        Object {
          "description": "MICROSOFT CORP",
          "displaySymbol": "MSFT",
          "symbol": "MSFT",
        },
      ]
    }
  />
  <div
    className="stock-detail"
  >
    <header
      className="stock-detail-header"
    >
      <span
        className="ui-text ui-text--body ui-text--bold stock-detail-symbol"
      >
        AAPL
      </span>
      <span
        className="ui-text ui-text--body stock-detail-description"
      >
        APPLE INC
      </span>
      <button
        className="ui-btn ui-btn--with-icon stock-detail-close-btn"
        onClick={[Function]}
        type="button"
      >
        <i
          className="ui-icon ui-icon-times-circle fas fa-times-circle fa-2x"
        />
        <span
          className="ui-text ui-text--accessible"
        >
          Close detail view
        </span>
      </button>
    </header>
    <div
      className="stock-detail-content"
    >
      <div
        className="stock-detail-value"
      >
        <span
          className="ui-text ui-text--body stock-detail-price"
        >
          
        </span>
        <span
          className="ui-text ui-text--body stock-detail-difference stock-detail-difference--down"
        />
      </div>
      <StockChart
        stock={
          Object {
            "description": "APPLE INC",
            "displaySymbol": "AAPL",
            "symbol": "AAPL",
          }
        }
      />
      <StockNews
        category="AAPL"
        displayCount={5}
      />
    </div>
  </div>
  <StockNews
    draggable={true}
  />
</div>
`;
